### PR TITLE
Add simple progressbar to publish dialog

### PR DESCRIPTION
**Pull request by TheCodez**
Currently has 3 progress states: Data file, Image file, Image optimization

### CODE REVIEW

File: SpriteSheetPacker/PublishSpriteSheet.cpp [hunk 1]
```diff
@@ -39,7 +39,7 @@ void PublishSpriteSheet::addSpriteSheet(const SpriteAtlas &atlas, const QString
     _fileNames.append(fileName);
 }
 
-bool PublishSpriteSheet::publish(const QString& format, const QString& optMode, int optLevel, bool errorMessage) {
+bool PublishSpriteSheet::publish(const QString& format, const QString& optMode, int optLevel, bool errorMessage, PublishStatusDialog* dialog) {
```
Comment (amakaseev) [inline SpriteSheetPacker/PublishSpriteSheet.cpp]: call calback and remove PublishStatusDialog* dialog
plz void (*Callback)(int currentStep, int allSteps)

and call cb(0, n) cb(1, n) .....

publish(......, [](int a, int b)) {
  publishStatusDialog->progresBar(...)
}

Comment (TheCodez) [inline SpriteSheetPacker/PublishSpriteSheet.cpp]: good idea :+1:

Comment (amakaseev) [inline SpriteSheetPacker/PublishSpriteSheet.cpp]: more nuance ...

max steps on progress bar = scaleVariant*allSteps...

File: SpriteSheetPacker/MainWindow.cpp [hunk 2]
```diff
@@ -606,27 +606,24 @@ void MainWindow::on_actionPublish_triggered() {
...
-        QEventLoop loop;
-        connect(&publishStatusDialog, SIGNAL(finished(int)), &loop, SLOT(quit()));
-        loop.exec();
-        qDebug() << "Finish publish";
-    }
+        } else {
+            QObject::connect(publisher, &PublishSpriteSheet::onCompleted, [publishStatusDialog, publisher] () {
+                publishStatusDialog->log(QString("Publishing is finished."), Qt::blue);
+                publishStatusDialog->complete();
+                publishStatusDialog->progressBar()->setValue(3);
```
Comment (amakaseev) [inline SpriteSheetPacker/MainWindow.cpp]: if complete - auto close
crash?

Comment (TheCodez) [inline SpriteSheetPacker/MainWindow.cpp]: haven't had any crash with auto close yet. But I will finish the implementation tomorrow.

File: SpriteSheetPacker/MainWindow.cpp [hunk 3]
```diff
@@ -606,27 +606,24 @@ void MainWindow::on_actionPublish_triggered() {
...
-        // This loop will wait for the window is finished
-        QEventLoop loop;
-        connect(&publishStatusDialog, SIGNAL(finished(int)), &loop, SLOT(quit()));
-        loop.exec();
-        qDebug() << "Finish publish";
-    }
+        } else {
+            QObject::connect(publisher, &PublishSpriteSheet::onCompleted, [publishStatusDialog, publisher] () {
+                publishStatusDialog->log(QString("Publishing is finished."), Qt::blue);
+                publishStatusDialog->complete();
```
Comment (amakaseev) [inline SpriteSheetPacker/MainWindow.cpp]: I understand correctly, you are again banned pablish close dialogue while there is optimization?

Comment (TheCodez) [inline SpriteSheetPacker/MainWindow.cpp]: Correct, the only button to close the dialog is the complete button that can only be pressed when optimization finished. It's probably a good idea to add a cancel button later.

Comment (amakaseev) [inline SpriteSheetPacker/MainWindow.cpp]: It is better to bring back as it was...
Optimization in thread is made to allow the user to continue working

and progress bar make in status bar, as I wrote earlier

**Comment by TheCodez** [comment]
So adding a progress bar in the area where the zoom slider is.

**Comment by amakaseev** [comment]
i found this [sample](http://www.bogotobogo.com/Qt/Qt5_QStatusBar.php)

**Comment by TheCodez** [comment]
is the progressbar in the status bar just the progress of optimization or also image and file writing?

**Comment by amakaseev** [comment]
You have added a stage of progress ..
but initially I thought only about the progress of optimization as the other stages take place in the publish dialog

**Comment by amakaseev** [comment]
it would be great if processes like this:
I opened the first project file - pressed publish
opened next project - publish
and so on...
a progress bar shows the overall progress of optimization that you could observe when they are ready all the resources
but maybe it's too much?

**Comment by TheCodez** [comment]
To be honest, I think it would be the best, to have the optimization in the publish dialog, but allow the user to cancel that in case he realizes he wants to use a different setting. When I press publish I expect to have all my resources ready when the dialog says finished, Texture packer does it the same way and I think it's a good concept.
However, you are the owner so you decide.

**Comment by amakaseev** [comment]
so it makes sense ...
it is necessary to pry as implemented in TexturtPacker

**Comment by TheCodez** [comment]
Not what I said :smile: I just like it that way more, but If you think that's a better idea than I'm going to change that.

**Comment by amakaseev** [comment]
I apologize if I misled you
Now analyze your version seems logical

**Comment by TheCodez** [comment]
What's also possible is make the whole publish process a background task and then show a messagebox when finished.

**Comment by amakaseev** [comment]
I just looked image optimization should not be blocking the task ...
since more than 2048 images optimized few minutes :(

**Comment by TheCodez** [comment]
Maybe like in eclipse you have a button "Run in background" that runs the optimization in background like you said

**Comment by amakaseev** [comment]
in the previous version it was, there was only a progress bar and a message about the end of the optimization suddenly pop up

**Comment by amakaseev** [comment]
It may be worth adding the cancel button next to the progress bar in the status bar?
and temporarily lock publish button that the user would not press her again until all optimization?

**Comment by amakaseev** [comment]
I think everyone was nice to expect when working application is not blocked

**Comment by TheCodez** [comment]
Yeah agree. With the publish button disabled then his shouldn't be a problem.

**Comment by TheCodez** [comment]
So it should look like this and only make it visible when publishing.

image

**Comment by amakaseev** [comment]
that's wonderful

**Comment by TheCodez** [comment]
Ok I'll start the implementation then.

**Comment by amakaseev** [comment]
I found this [example](http://blackberry.github.io/Qt2Cascades-Samples/docs/qtconcurrent-progressdialog.html)
I hope it will help you, thank you very much.

**Comment by TheCodez** [comment]
Yeah, I will use parts of that. Ideally i should be using QConcurrent::map but that didn't quite work so I'm using a for loop wit run for the moment.

**Comment by TheCodez** [comment]
Is it ok if I implement the cancel button in a later pr. Have the base implementation finished.

**Comment by TheCodez** [comment]
I feel like the publish status dialog is kinda useless if there is a progressbar in the statusbar

**Comment by amakaseev** [comment]
publish status dialog needs for show logs
all proceses in main thred only optimize png in other thread
at the moment

**Pull request closed without merging**